db/badgerdb/badger.go

```diff
@@ -0,0 +1,145 @@
+package badgerdb
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/dgraph-io/badger/v3"
+	"go.vocdoni.io/dvote/db"
+)
+
+// MEMTABLESIZE defines the BadgerDB maximum size in bytes for memtable table.
+// The default is 64<<20 (64MB), this does not pre-allocate enough memory for
+// big Txs, that's why we use 128<<20.
+const MEMTABLESIZE = 128 << 20
+
+type ReadTx struct {
+	tx *badger.Txn
+}
+
+type WriteTx struct {
+	ReadTx
+	tx *badger.Txn
+}
+
+func (tx ReadTx) Get(k []byte) ([]byte, error) {
+	item, err := tx.tx.Get(k)
+	if err == badger.ErrKeyNotFound {
+		return nil, db.ErrKeyNotFound
+	}
+	if err != nil {
+		return nil, err
+	}
+	return item.ValueCopy(nil)
+}
+func (tx ReadTx) Discard() {
+	tx.tx.Discard()
+}
+
+func (tx WriteTx) Get(k []byte) ([]byte, error) {
+	item, err := tx.tx.Get(k)
+	if err == badger.ErrKeyNotFound {
+		return nil, db.ErrKeyNotFound
+	}
+	if err != nil {
+		return nil, err
+	}
+	return item.ValueCopy(nil)
+}
+func (tx WriteTx) Set(k, v []byte) error {
+	if err := tx.tx.Set(k, v); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (tx WriteTx) Del(k []byte) error {
+	return tx.tx.Delete(k)
+}
+
+func (tx WriteTx) Commit() error {
+	return tx.tx.Commit()
+}
+func (tx WriteTx) Discard() {
+	tx.tx.Discard()
+}
+
+// check that Tx interfaces are matched
+var _ db.ReadTx = (*ReadTx)(nil)
+var _ db.ReadTx = (*WriteTx)(nil)
+var _ db.WriteTx = (*WriteTx)(nil)
+
+// BadgerDB implements db.Database interface
+type BadgerDB struct {
+	path string
+	db   *badger.DB
+}
+
+var _ db.Database = (*BadgerDB)(nil)
+
+func NewBadgerDB(path string) (*BadgerDB, error) {
+	if err := os.MkdirAll(path, os.ModePerm); err != nil {
+		return nil, err
+	}
+	opts := badger.DefaultOptions(path).
+		WithLogger(nil).
+		// Do we want compression in the future?
+		// WithCompression(options.Snappy).
+		WithSyncWrites(false)
+
+	opts.MemTableSize = MEMTABLESIZE
+
+	db, err := badger.Open(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BadgerDB{
+		path: path,
+		db:   db,
+	}, nil
+}
+
+func (db *BadgerDB) ReadTx() db.ReadTx {
+	return ReadTx{
+		tx: db.db.NewTransaction(false),
+	}
+}
+func (db *BadgerDB) WriteTx() db.WriteTx {
+	return WriteTx{
+		tx: db.db.NewTransaction(true),
+	}
+}
+
+func (db *BadgerDB) Close() error {
+	return db.db.Close()
+}
+
+// stopIterations is used to stop from iterating in the case that the callback
+// method passed into Iterate requires it
+var stopIterations = fmt.Errorf("stop iterations")
```

Comment:
use the `err` prefix for global error vars please

Reply:
In this case, is not an 'error' as the concept of error, but an aux variable that is returned to indicate to stop iterating, but not necessary because of an error, it can be a desired behavior, like for example wanting to iterate until finding a key X, so when that key X is found it stops iterating by using this aux variable.
Anyway, I've updated the name of the variable to `errStopIterations` as suggested, but is important to have clear that is not used as an error. The description of the variable in the inline code comment should give a notion on how this 'error' is used (not as an error).